src/NumericalAlgorithms/LinearSolver/LinearSolver.hpp

```diff
@@ -0,0 +1,279 @@
+#include "Options/Auto.hpp"
+#include "Options/Options.hpp"
+#include "Parallel/CharmPupable.hpp"
+#include "Parallel/PupStlCpp17.hpp"
+#include "Utilities/FakeVirtual.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Registration.hpp"
+#include "Utilities/Requires.hpp"
+
+namespace LinearSolver::Serial {
+
+/// Indicates the linear solver uses no preconditioner. It may perform
+/// compile-time optimization for this case.
+struct NoPreconditioner {};
+
+/// Indicates the linear solver uses any other registered linear solver as
+/// preconditioner.
+struct AnyPreconditioner {};
+
+/// Registrars for linear solvers
+namespace Registrars {}
+
+/*!
+ * \brief Base class for serial linear solvers that supports factory-creation
+ * and nested preconditioning.
+ *
+ * To enable support for preconditioning in your derived linear solver class,
+ * pass any type that has a `solve` and a `reset` function as the
+ * `Preconditioner` template parameter. It can also be an abstract
+ * `LinearSolver` type, which means that any other linear solver can be used as
+ * preconditioner. Pass `NoPreconditioner` to disable support for
+ * preconditioning.
+ */
+template <typename LinearSolverRegistrars>
+class LinearSolver : public PUP::able {
+ protected:
+  /// \cond
+  LinearSolver() = default;
+  LinearSolver(const LinearSolver&) = default;
+  LinearSolver(LinearSolver&&) = default;
+  LinearSolver& operator=(const LinearSolver&) = default;
+  LinearSolver& operator=(LinearSolver&&) = default;
+  /// \endcond
+
+ public:
+  ~LinearSolver() override = default;
+
+  /// \cond
+  explicit LinearSolver(CkMigrateMessage* m) noexcept;
+  WRAPPED_PUPable_abstract(LinearSolver);  // NOLINT
+  /// \endcond
+
+  using creatable_classes = Registration::registrants<LinearSolverRegistrars>;
+
+  virtual std::unique_ptr<LinearSolver<LinearSolverRegistrars>> get_clone()
+      const = 0;
+
+  /*!
+   * \brief Solve the linear equation \f$Ax=b\f$ where \f$A\f$ is the
+   * `linear_operator` and \f$b\f$ is the `source`.
+   *
+   * - The (approximate) solution \f$x\f$ is returned in the
+   *   `initial_guess_in_solution_out` buffer, which also serves to provide an
+   *   initial guess for \f$x\f$. Not all solvers take the initial guess into
+   *   account, but expect the buffer is sized correctly.
+   * - The `linear_operator` must be an invocable that takes a `VarsType` as
+   *   const-ref argument and returns a `SourceType` by reference.
+   *
+   * Each solve may mutate the private state of the solver, for example to cache
+   * quantities to accelerate successive solves for the same operator. Invoke
+   * `reset` to discard these caches.
+   */
+  template <typename LinearOperator, typename VarsType, typename SourceType,
+            typename... Args>
+  Convergence::HasConverged solve(
+      gsl::not_null<VarsType*> initial_guess_in_solution_out,
+      const LinearOperator& linear_operator, const SourceType& source,
+      Args&&... args) const noexcept;
```

Comment:
For my understanding: if `solve` will in general mutate the state of the solver, then why is `solve` a `const` member function?

Reply:
It may only mutate _private_ state, e.g. resize buffers so they don't have to be re-allocated, or cache data to accelerate consecutive solves. None of this is public state, i.e. the parameters such as convergence criteria remain `const`. It only means that consecutive solves are faster. Sounds OK?